[FIX #46021] Allow passing classes to dom_id

goulvench opened the pull request:
### Motivation / Background

As mentioned in #46021, allowing `dom_id` to receive a class makes it behave like `dom_class`, saves a few keystrokes, and saves memory because Ruby doesn't need to instantiate an entire object (which could have `after_initialize` callbacks, etc).

The last commit allows receiving non-ActiveModel classes, in case this change is deemed unnecessary. 

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [X] There are no typos in commit messages and comments.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Feature branch is up-to-date with `main` (if not - rebase it).
* [X] Pull request only contains one commit for bug fixes and small features. If it's a larger feature, multiple commits are permitted but must be descriptive. (see comment above regarding non-ActiveModel)
* [X] Tests are added if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [X] PR is not in a draft state.
* [X] CI is passing. (extra whitespace in the changelog, I'm on it)

goulvench committed:
Allow passing a class to `dom_id`

You no longer need to call `new` when passing a class to `dom_id`.
This makes `dom_id` behave like `dom_class` in this regard.
Apart from saving a few keystrokes, it prevents Ruby from needing to
instantiate a whole new object just to generate a string.

Before:
```ruby
dom_id(Post) # NoMethodError: undefined method `to_key' for Post:Class
```

After:
```ruby
dom_id(Post) # "new_post"
```

You can still call `dom_id(Post.new)`.

rafaelfranca commented:
We can delete the last commit. Action View requires the Active Model API to to be implemented in the objects.

goulvench commented:
Thanks for the quick response. I've removed the last commit which allowed `dom_id` to work with non-ActiveModel, and squashed all commits. Removed the extra whitespace as well.

Should this be backported to the Rails 6 branch?

eileencodes commented:
> Should this be backported to the Rails 6 branch?

6-1-stable and 6-0-stable are not supported for bug fixes only the 7-0-stable branch is. Features can only go into `main` and this is more a feature than a bug fix.

eileencodes commented on actionview/CHANGELOG.md:
```diff
@@ -1,3 +1,27 @@
+    Apart from saving a few keystrokes, it prevents Ruby from needing
+    to instantiate a whole new object just to generate a string.
+
+    Before:
```
```suggestion
    Before:
    
```

eileencodes commented on actionview/CHANGELOG.md:
````diff
@@ -1,3 +1,27 @@
+    dom_id(Post)        # => NoMethodError: undefined method `to_key' for Post:Class
+    ```
+
+    After:
````
```suggestion

    After:
```

eileencodes commented on actionview/CHANGELOG.md:
```diff
@@ -1,3 +1,27 @@
+    dom_class(Post.new) # => "new_post"
+    dom_class(Post)     # => "new_post"
+    dom_id(Post.new)    # => "new_post"
+    dom_id(Post)        # => "new_post"
```
We really only need the last example in this changelog since the behavior only changed for passing a class to `dom_id`.

eileencodes commented on actionview/lib/action_view/record_identifier.rb:
```diff
@@ -80,17 +80,17 @@ def dom_class(record_or_class, prefix = nil)
     #
     #   dom_id(Post.find(45))       # => "post_45"
-    #   dom_id(Post.new)            # => "new_post"
+    #   dom_id(Post)                # => "new_post"
```
The previous example still works so we should leave both examples in the docs.

goulvench replied:
I left only the new syntax because that's what was documented for dom_class, and to encourage people to use the new one. It seems implicit to me that the old one still works, but I won't argue if you think it's best to leave both. Or should we add the `dom_class(Post.new)` example syntax too?

eileencodes commented on actionview/lib/action_view/record_identifier.rb:
```diff
@@ -103,8 +103,9 @@ def dom_id(record, prefix = nil)
-      key = convert_to_model(record).to_key
+    def record_key_for_dom_id(record_or_class) # :doc:
+      model = convert_to_model(record_or_class)
+      key = model.respond_to?(:to_key) ? model.to_key : nil
```
I'm on the fence about this method implementation since it can only work for a model and never for a class. I feel we shouldn't even call this method if we have a class rather than having to twice check our value and return `nil` if it's a class.

goulvench replied:
I've updated the logic in `dom_id` following your feedback, is it better now?

```ruby
def dom_id(record_or_class, prefix = nil)
  record_id = record_key_for_dom_id(record_or_class) unless record_or_class.is_a?(Class)
  if record_id
    "#{dom_class(record_or_class, prefix)}#{JOIN}#{record_id}"
  else
    dom_class(record_or_class, prefix || NEW)
  end
end
```

goulvench commented:
Thanks @rafaelfranca and @eileencodes for the feedback and merging my first contribution to Rails! 🥳
